File: src/main/java/com/pauljoda/modularsystems/crusher/tiles/TileCrusherCore.java
package com.pauljoda.modularsystems.crusher.tiles;

import com.pauljoda.modularsystems.core.blocks.BlockDummy;
import com.pauljoda.modularsystems.core.functions.BlockCountFunction;
import com.pauljoda.modularsystems.core.managers.BlockManager;
import com.pauljoda.modularsystems.core.registries.BlockValueRegistry;
import com.pauljoda.modularsystems.core.registries.CrusherRecipeRegistry;
import com.pauljoda.modularsystems.core.registries.FurnaceBannedBlocks;
import com.pauljoda.modularsystems.core.tiles.AbstractCore;
import com.pauljoda.modularsystems.crusher.blocks.BlockCrusherCore;
import com.pauljoda.modularsystems.crusher.container.ContainerCrusher;
import com.pauljoda.modularsystems.crusher.gui.GuiCrusher;
import com.teambr.bookshelf.common.tiles.IOpensGui;
import com.teambr.bookshelf.helpers.BlockHelper;
import net.minecraft.block.Block;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.inventory.Container;
import net.minecraft.item.ItemStack;
import net.minecraft.world.World;

public class TileCrusherCore extends AbstractCore implements IOpensGui {

    @Override
    protected void updateBlockState(boolean positiveBurnTime, World world, int x, int y, int z) {
        BlockCrusherCore.updateCrusherBlockState(positiveBurnTime, world, x, y, z);
    }

    @Override
    protected ItemStack recipe(ItemStack is) {
        return is == null ? null : CrusherRecipeRegistry.INSTANCE.getOutput(is);
    }

    @Override
    @SuppressWarnings("all")
    protected boolean isBlockBanned(Block block, int meta) {
        return block instanceof BlockDummy ? false : FurnaceBannedBlocks.INSTANCE.isBlockBanned(block, meta) || FurnaceBannedBlocks.isBadBlockFromBlock(block);
    }

    @Override
    protected void generateValues(BlockCountFunction function) {
        for(String i : function.getBlockIds()) {
            if (BlockValueRegistry.INSTANCE.isBlockRegistered(BlockHelper.getBlockFromString(i).getFirst(), BlockHelper.getBlockFromString(i).getSecond())) {
                values.addToSpeed(BlockValueRegistry.INSTANCE.getSpeedValue(BlockHelper.getBlockFromString(i).getFirst(), BlockHelper.getBlockFromString(i).getSecond(), function.getBlockCount(BlockHelper.getBlockFromString(i).getFirst(), BlockHelper.getBlockFromString(i).getSecond())));
                values.addToEfficiency(BlockValueRegistry.INSTANCE.getEfficiencyValue(BlockHelper.getBlockFromString(i).getFirst(), BlockHelper.getBlockFromString(i).getSecond(), function.getBlockCount(BlockHelper.getBlockFromString(i).getFirst(), BlockHelper.getBlockFromString(i).getSecond())));
                values.addToMultiplicity(BlockValueRegistry.INSTANCE.getMultiplicityValue(BlockHelper.getBlockFromString(i).getFirst(), BlockHelper.getBlockFromString(i).getSecond(), function.getBlockCount(BlockHelper.getBlockFromString(i).getFirst(), BlockHelper.getBlockFromString(i).getSecond())));
            }
        }

        for(String i : function.getMaterialStrings()) {
            if(BlockValueRegistry.INSTANCE.isMaterialRegistered(i)) {
                values.addToSpeed(BlockValueRegistry.INSTANCE.getSpeedValueMaterial(i, function.getMaterialCount(i)));
                values.addToEfficiency(BlockValueRegistry.INSTANCE.getEfficiencyValueMaterial(i, function.getMaterialCount(i)));
                values.addToMultiplicity(BlockValueRegistry.INSTANCE.getMultiplicityValueMaterial(i, function.getMaterialCount(i)));
            }
        }
    }

    @Override
    protected Block getOnBlock() {
        return BlockManager.crusherCoreActive;
    }

    @Override
    public int getRedstoneOutput() {
        return Container.calcRedstoneFromInventory(this);
    }

    @Override
    public Object getServerGuiElement(int ID, EntityPlayer player, World world, int x, int y, int z) {
        return new ContainerCrusher(player.inventory, this);
    }

    @Override
    public Object getClientGuiElement(int ID, EntityPlayer player, World world, int x, int y, int z) {
        return new GuiCrusher(player.inventory, this);
    }
}

File: src/main/scala/com/teambr/modularsystems/crusher/tiles/TileCrusherCore.scala
package com.teambr.modularsystems.crusher.tiles

import com.teambr.bookshelf.collections.Location
import com.teambr.bookshelf.helper.BlockHelper
import com.teambr.bookshelf.util.InventoryUtils
import com.teambr.modularsystems.core.common.blocks.BlockProxy
import com.teambr.modularsystems.core.common.tiles.AbstractCore
import com.teambr.modularsystems.core.functions.BlockCountFunction
import com.teambr.modularsystems.core.managers.BlockManager
import com.teambr.modularsystems.core.registries.{BlockValueRegistry, CrusherRecipeRegistry, FurnaceBannedBlocks}
import net.minecraft.block.Block
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumFacing

import scala.util.Random

/**
 * This file was created for Modular-Systems
 *
 * Modular-Systems is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Dyonovan
 * @since August 08, 2015
 */
class TileCrusherCore extends AbstractCore {

    /*******************************************************************************************************************
      ************************************************  Furnace Methods  ***********************************************
      ******************************************************************************************************************/

    override def smeltItem() {
        val smeltCount: (Int, Int) = smeltCountAndSmeltSize
        if (smeltCount != null && smeltCount._2 > 0) {
            var recipeResult: ItemStack = recipe(getStackInSlot(0))
            decrStackSize(0, smeltCount._1)
            //getStackInSlot(0).stackSize -= smeltCount._1
            if (getStackInSlot(1) == null) {
                recipeResult = recipeResult.copy
                recipeResult.stackSize = smeltCount._2
                setInventorySlotContents(1, recipeResult)
            }
            else {
                getStackInSlot(1).stackSize += smeltCount._2
            }
            extraOutput(getStackInSlot(0), smeltCount._1)
        }
    }

    def extraOutput(input: ItemStack, count: Int): Unit = {
        if (getCrusherExtraCount > 0) {
            val recipeResult = recipeCrusher(input)
            if (recipeResult != null && recipeResult._2 != null) {
                var extraCount = 0
                for (i <- 0 until count) {
                    val random = Random.nextInt(100)
                    if (getCrusherExtraCount >= random) {
                        extraCount += 1
                    }
                }
                val extra = recipeResult._2.copy

                if (getStackInSlot(2) == null && extraCount > 0) {
                    if (extraCount > extra.getMaxStackSize)
                        extra.stackSize = extra.getMaxStackSize
                    else
                        extra.stackSize = extraCount
                    setInventorySlotContents(2, extra)
                } else if (extraCount > 0 && getStackInSlot(2).isItemEqual(extra)) {
                    if (extraCount + getStackInSlot(2).stackSize > extra.getMaxStackSize)
                        getStackInSlot(2).stackSize = extra.getMaxStackSize
                    else
                        getStackInSlot(2).stackSize += extraCount
                }
            }
        }
    }

    def getCrusherExtraCount: Int = {
        val coords = new Location(corners._1).getAllWithinBounds(new Location(corners._2), includeInner = false, includeOuter = true)
        var count = 0
        for (i <- 0 until coords.size()) {
            val coord = coords.get(i)
            val block: Block = worldObj.getBlockState(coord.asBlockPos).getBlock
            if (block != null) {
                if (block == BlockManager.crusherExpansion) count += 1
            }
        }
        if (count * 10 > 100) 100 else count * 10
    }

    /**
     * Get the output of the recipe
 *
     * @param stack The input
     * @return The output
     */
    override def recipe(stack: ItemStack): ItemStack = if (recipeCrusher(stack) == null) null else recipeCrusher(stack)._1

    def recipeCrusher(stack: ItemStack): (ItemStack, ItemStack) = {
        CrusherRecipeRegistry.getOutput(stack).orNull
    }

    /*******************************************************************************************************************
      ************************************************  Core Methods  **************************************************
      ******************************************************************************************************************/

    /**
     * Take the blocks in this structure and generate the speed etc values
 *
     * @param function The blocks count function
     */
    override def generateValues(function: BlockCountFunction): Unit = {
        for (i <- function.getBlockIds) {
            val t = (BlockHelper.getBlockFromString(i)._1, BlockHelper.getBlockFromString(i)._2)

            if (BlockValueRegistry.isBlockRegistered(t._1, t._2)) {
                values.speed += BlockValueRegistry.getSpeedValue(t._1, t._2, function.getBlockCount(t._1, t._2))
                values.efficiency += BlockValueRegistry.getEfficiencyValue(t._1, t._2, function.getBlockCount(t._1, t._2))
                values.multiplicity += BlockValueRegistry.getMultiplicityValue(t._1, t._2, function.getBlockCount(t._1, t._2))
            }
        }

        for (i <- function.getMaterialStrings) {
            if (BlockValueRegistry.isMaterialRegistered(i)) {
                values.speed += BlockValueRegistry.getSpeedValueMaterial(i, function.getMaterialCount(i))
                values.efficiency += BlockValueRegistry.getEfficiencyValueMaterial(i, function.getMaterialCount(i))
                values.multiplicity += BlockValueRegistry.getMultiplicityValueMaterial(i, function.getMaterialCount(i))
            }
        }
    }

    /**
     * Check if this blocks is not allowed in the structure
 *
     * @param block The blocks to check
     * @param meta The meta data of said blocks
     * @return True if it is banned
     */
    override def isBlockBanned(block: Block, meta: Int): Boolean = {
        if (block.isInstanceOf[BlockProxy])
            false
        else
            FurnaceBannedBlocks.isBlockBanned(block, meta)
    }

    /**
     * Used to output the redstone single from this structure
     *
     * Use a range from 0 - 16.
     *
     * 0 Usually means that there is nothing in the tile, so take that for lowest level. Like the generator has no energy while
     * 16 is usually the flip side of that. Output 16 when it is totally full and not less
     *
     * @return int range 0 - 16
     */
    override def getRedstoneOutput: Int = InventoryUtils.calcRedstoneFromInventory(this)

    /*******************************************************************************************************************
      ************************************************* Inventory methods ***********************************************
      *******************************************************************************************************************/

    override def initialSize: Int = 3

     /**
     * Returns true if automation can extract the given item in the given slot from the given side. Args: slot, item,
     * side
     */
    override def canExtractItem(index : Int, stack : ItemStack, direction : EnumFacing) : Boolean = {index == 1 || index == 2}

    override def getSlotsForFace(side : EnumFacing) : Array[Int] = {
        Array[Int](0, 1, 2)
    }
}
